h/security/policy/_cookie.py

```diff
@@ -14,13 +14,46 @@ class CookiePolicy:
     straps the login for the client (when the popup shows).
     """
 
-    def __init__(self, cookie: SignedCookieProfile, helper: AuthTicketCookieHelper):
-        self.cookie = cookie
+    def __init__(
+        self,
+        html_authcookie: SignedCookieProfile,
+        api_authcookie: SignedCookieProfile,
+        helper: AuthTicketCookieHelper,
+    ):
+        self.html_authcookie = html_authcookie
+        self.api_authcookie = api_authcookie
         self.helper = helper
 
     def identity(self, request):
         self.helper.add_vary_by_cookie(request)
-        return self.helper.identity(self.cookie, request)
+        identity = self.helper.identity(self.html_authcookie, request)
+
+        # If a request was successfully authenticated using the HTML auth
+        # cookie and that request did *not* also contain the API auth cookie,
+        # then add the API auth cookie to the user's browser.
```

Comment:
This should never happen:

Both cookies are always set at the same time (at login) and deleted at the same time (on logout) and have the same `max-age`, and for existing logged-in users who have the HTML cookie but not the API cookie this PR contains transitional code to detect that situation and issue an API cookie that will have a `max-age` beyond that of the existing HTML cookie.

So the browser should always have either both cookies or neither.

This can't currently be broken by a user going into their dev tools and deleting the API cookie but not the HTML one: the transitional code would detect that situation and issue a new API cookie. If https://github.com/hypothesis/h/pull/8865 was merged then the transitional code would no longer be present to save us in that situation, so perhaps we should consider closing https://github.com/hypothesis/h/pull/8865 and making the transitional code permanent.

Having a browser extension that blocks the API cookie but not the HTML one would still be a problem, but that seems unlikely to me particularly because they would have to specifically block or delete the API cookie but not the HTML one.

Nonetheless, I've artificially created this condition (when the browser has the HTML cookie but not the API one). On the create-group page this is what happens:

![image](https://github.com/user-attachments/assets/e1f21997-b26e-44cc-862d-d5a3432300dd)

The response that the frontend receives to its unauthenticated create-group API request is **404 Not Found** with this JSON body:

```json
{
  "status": "failure",
  "reason": "Either the resource you requested doesn't exist, or you are not currently authorized to see it."
}
```

I've created an issue to track this as potential follow-up work: https://github.com/hypothesis/h/issues/8881

Reply:
Added a commit to remove comments about the transitional API cookie issuing code being temporary, in case we decide to make it permanent: https://github.com/hypothesis/h/pull/8861/commits/d7ef927f0ca3fa517f6a02d689454d0ba0d1fbd5